### PR TITLE
Refactor/operation worker execution flow

### PR DESCRIPTION
**Pull request by guilhermocc**
## What ❓ 
Refactor operation execution worker code, isolating the core execution logic into separate functions.

## Why 🤔 
Currently, our core execution logic resides inside a loop and is not isolated into a function, making it hard to expand operation execution worker logic in the future.

### CODE REVIEW

File: internal/core/workers/operation_execution_worker/operation_execution_worker.go [hunk 1]
```diff
@@ -76,125 +79,158 @@ func (w *OperationExecutionWorker) Start(ctx context.Context) error {
...
-			loopLogger.Warn("operation definition has no executor")
-
-			w.evictOperation(ctx, loopLogger, op)
-			reportOperationEvicted(w.scheduler.Game, w.scheduler.Name, op.DefinitionName, LabelNoOperationExecutorFound)
-
-			continue
+		err = w.executeOperation(op, def)
+		if err != nil {
+			switch {
+			case errors.Is(err, workererrors.ErrStartOperationFailed) ||
```
Comment (guilhermocc) [inline internal/core/workers/operation_execution_worker/operation_execution_worker.go:89]: Now it is easier to decide if we should stop the worker in those scenarios. Initially, I made just a refactor, didn't change the logic at all... Now it is a chance to change this behavior, @arthur29 @caiordjesus what do you guys think? should we continue to stop the worker and let the worker manage to recreate it in those scenarios? Do you guys think it is ok to change this behavior is this PR?

File: internal/core/workers/operation_execution_worker/operation_execution_worker.go [hunk 2]
```diff
@@ -76,125 +79,158 @@ func (w *OperationExecutionWorker) Start(ctx context.Context) error {
...
+				errors.Is(err, workererrors.ErrGrantLeaseFailed) ||
+				errors.Is(err, workererrors.ErrOperationWithInvalidStatus):
+				return err
+			}
```
Comment (arthur29) [inline internal/core/workers/operation_execution_worker/operation_execution_worker.go:93]: What do you think to refactor this code to be more simple like:
```
		if err != nil && shouldFinishWorker(err)  {
			return err
		}
	}
}

func shouldFinishWorker(err error) bool {
	return errors.Is(err, workererrors.ErrStartOperationFailed) ||
	errors.Is(err, workererrors.ErrGrantLeaseFailed) ||
	errors.Is(err, workererrors.ErrOperationWithInvalidStatus)
}
```

File: internal/core/workers/operation_execution_worker/operation_execution_worker.go [hunk 3]
```diff
@@ -76,125 +79,158 @@ func (w *OperationExecutionWorker) Start(ctx context.Context) error {
...
+		zap.String(logs.LogFieldOperationID, op.ID),
+		zap.String(logs.LogFieldOperationDefinition, def.Name()),
+	)
 
-		loopLogger.Info("Starting operation")
+	if op.Status != operation.StatusPending {
+		loopLogger.Warn("operation is at an invalid status to proceed")
 
-		w.operationManager.AppendOperationEventToExecutionHistory(ctx, op, "Starting operation")
+		return workererrors.NewErrOperationWithInvalidStatus("operation is at and invalid status to proceed")
```
Comment (arthur29) [inline internal/core/workers/operation_execution_worker/operation_execution_worker.go:107]:
```suggestion
		return workererrors.NewErrOperationWithInvalidStatus("operation is at an invalid status to proceed")
```

File: internal/core/workers/operation_execution_worker/operation_execution_worker.go [hunk 4]
```diff
@@ -76,125 +79,158 @@ func (w *OperationExecutionWorker) Start(ctx context.Context) error {
...
+
+		return workererrors.NewErrStartOperationFailed("failed to start operation \"%s\" for the scheduler \"%s\"", op.ID, op.SchedulerName), nil
+	}
+
+	w.operationManager.StartLeaseRenewGoRoutine(operationContext, op)
+
+	executionErr = w.executeCollectingLatencyMetrics(op.DefinitionName, func() operations.ExecutionError {
+		return executor.Execute(operationContext, op, def)
+	})
+	return err, executionErr
```
Comment (arthur29) [inline internal/core/workers/operation_execution_worker/operation_execution_worker.go:215]: Hey @guilhermocc why had you done this one? It is kinda of weird.

File: internal/core/workers/operation_execution_worker/operation_execution_worker.go [hunk 5]
```diff
@@ -76,125 +79,158 @@ func (w *OperationExecutionWorker) Start(ctx context.Context) error {
...
-			reportOperationExecutionWorkerFailed(w.scheduler.Game, w.scheduler.Name, LabelStartOperationFailed)
-			operationCancellationFunction()
+	executor, hasExecutor := w.executorsByName[def.Name()]
+	if w.shouldEvictOperation(op, def, hasExecutor, loopLogger) {
+		return nil
+	}
 
-			op.Status = operation.StatusError
-			w.operationManager.AppendOperationEventToExecutionHistory(ctx, op, fmt.Sprintf("Failed to grant lease to operation, reason: %s", err.Error()))
+	err, executionErr := w.startAndExecuteOperationWithLease(op, def, loopLogger, executor)
```
Comment (arthur29) [inline internal/core/workers/operation_execution_worker/operation_execution_worker.go:115]: What do you think to unify these errors in only one?

File: internal/core/workers/operation_execution_worker/operation_execution_worker.go [hunk 6]
```diff
@@ -76,125 +79,158 @@ func (w *OperationExecutionWorker) Start(ctx context.Context) error {
...
+	if err != nil {
+		return err
+	}
 
-			err = w.operationManager.FinishOperation(ctx, op)
-			if err != nil {
-				loopLogger.Error("failed to finish operation", zap.Error(err))
-			}
+	w.treatExecutionError(op, def, executionErr, loopLogger, executor)
```
Comment (arthur29) [inline internal/core/workers/operation_execution_worker/operation_execution_worker.go:120]: This function is receiving too many arguments, what do you think to split them into two functions, the first one to handle the error and the second one to execute the rollback, in this way I think that you could decrease the number of arguments to these functions. Last but not least comment is to move these functions into the `if err != nil` so you could remove that first if in the `treatExecutionError`.

**Review by arthur29 — CHANGES_REQUESTED** [review]

File: internal/core/workers/operation_execution_worker/operation_execution_worker.go (hunk 5 above)
Comment (guilhermocc) [reply internal/core/workers/operation_execution_worker/operation_execution_worker.go:115]: err and executionErr are different errors, not sure how we would unify them

**Review by arthur29 — APPROVED** [review]
LGTM

**Pull request merged**